### ISSUE TITLE
FEAT: Adaptive Knowledge Dimensions — domain-detected extension of UniversalKnowledge for project-specific intelligence (Solidity gas costs, firmware stack depth, ML experiment lineage)

### ISSUE THREAD

**Issue by nateschmiedehaus** [nateschmiedehaus/LiBrainian#688]
# FEAT: Adaptive Knowledge Dimensions — domain-detected extension of UniversalKnowledge for project-specific intelligence (Solidity gas costs, firmware stack depth, ML experiment lineage)

## Problem Statement

LiBrainian's `UniversalKnowledge` schema is designed for general-purpose software but optimized in practice for TypeScript/web projects. A Solidity contract author needs `gasComplexity` and `reentrancyRisk`. A firmware engineer needs `stackDepth` and `MISRA_compliance`. An ML engineer needs `experimentLineage` and `reproScore`. A medical device team needs `regulatoryTraceability`. Today, all projects receive the same fixed knowledge schema with the same LLM extraction prompts — the right dimensions for the project's domain are absent. LiBrainian is genuinely less useful for non-web projects as a result.

## Distinct From Existing Issues

- **#411 (Role-Aware Context):** Serves different *slices* (developer/architect/security engineer) of the **same fixed knowledge dimensions**. The dimensions themselves do not change — only their ordering and prominence. Adaptive Knowledge Dimensions changes **which dimensions exist** based on the detected project domain. These are orthogonal.

### What the audit found

**src/knowledge/universal_types.ts** — `UniversalKnowledge` is a **completely closed schema** with 14 named top-level sections: `semantics`, `contract`, `relationships`, `quality`, `security`, `runtime`, `testing`, `history`, `ownership`, `rationale`, `context`, `traceability`, `meta`, and optional `agentic`. There is **no extension point** (`Record<string, unknown>` or otherwise). Every field is typed and named. Adding a Solidity `gasComplexity` field requires modifying the core schema or adding an escape hatch.

**src/engines/domain_registry.ts** — `COMPLETE_DOMAIN_TAXONOMY` is a large static object with 300+ domain IDs covering engineering, design, content, product, management, quality, industries, and meta capabilities. `getDomainExpertise()` returns `null` for all domains ("For now, returns null for unimplemented domains"). The taxonomy is **reference data only** with no live consumers in the extraction pipeline.

**src/knowledge/registry/domain_registry.ts** — A separate `DomainRegistry` class with built-in domains (`identity`, `semantics`, `structure`, etc.) and support for custom domains loaded from config. Manages freshness and access patterns. **Not wired into `UniversalKnowledgeGenerator`** — the generator does not call or reference this registry.

**src/knowledge/generator.ts** — `UniversalKnowledgeGenerator.generateForFunction()` calls extractors in sequence: `extractSemanticsWithLLM`, `extractQuality`, `extractSecurityWithLLM`, `extractTesting`, `extractHistory`, `extractOwnership`, `extractRationaleWithLLM`, `extractTraceability`, `extractRelationships`. None of these extractors receive a `projectType`, `detectedDomain`, or domain context of any kind.

**src/knowledge/extractors/semantics.ts** — `SemanticsInput` contains `name`, `signature`, `content`, `docstring`, `existingPurpose`, `filePath`. The LLM prompt uses `filePath` and function name but has **no domain-specific injection**. A `.sol` file gets the same prompt as a `.ts` file.

**src/orchestrator/unified_init.ts** — `projectType` is computed at line 433: `constructableConfig.analysis.projectTypes[0]?.type ?? 'unknown'` and logged at line 667. It is **not passed to the knowledge generator or any extractor**.

**Conclusion:** The domain taxonomy exists as dead reference data. The knowledge registry exists but is not wired into extraction. `projectType` is computed but not consumed by the knowledge pipeline. The `UniversalKnowledge` schema has no extension point. All extractors are domain-blind.

### Domain signals for detection

At indexing time, file extensions and import patterns provide reliable domain signals:

| Signal | Domain | Dimensions to add |
|--------|--------|-------------------|
| `*.sol` files | Blockchain/Solidity | `gasComplexity`, `reentrancyRisk`, `storageSlotLayout`, `upgradeCompatibility`, `eventEmissions` |
| `*.tf`, `*.hcl` | Infrastructure (Terraform) | `destroyRisk`, `stateDrift`, `crossModuleRefs`, `providerConstraints`, `resourceCount` |
| `torch`, `tensorflow`, `sklearn` imports | ML/AI | `experimentLineage`, `reproScore`, `dataSchemaVersion`, `modelArtifacts`, `trainingConfig` |
| `MISRA`, bare-metal patterns, `volatile`, `ISR_` | Embedded/Firmware | `stackDepth`, `wcet`, `ISR_latency`, `heapAllocation`, `interruptSafety` |
| IEC 62304 comments, FDA-related patterns | Medical | `regulatoryTraceability`, `designHistoryRef`, `safetyClass`, `verificationStatus` |
| `contract`, `address`, `mapping(` patterns | Solidity (alternative) | Same as `*.sol` |

### Schema extension point

Add to `UniversalKnowledge`:

```typescript
// In universal_types.ts — new optional field
domainDimensions?: Record<string, DomainDimension>;

export interface DomainDimension {
  value: number | string | boolean | string[];
  unit?: string;                  // e.g., "gas units", "bytes", "ms"
  confidence: number;             // 0–1
  evidenceRefs: string[];         // What code patterns drove this value
  extractedBy: 'static' | 'llm'; // Which extractor produced it
  domain: string;                 // e.g., "solidity", "embedded", "ml"
}
```

This is an open extension point that avoids modifying the core named sections for each new domain. It allows domain-specific dimensions to appear alongside standard dimensions without breaking existing consumers.

### LLM extraction injection

The `SemanticsInput` type and prompt template in `semantics.ts` are domain-blind. When a domain is detected, inject domain context into the extraction prompt:

```
// Current prompt (all files):
"Provide a JSON response with purpose, domain, intent, mechanism, and complexity."

// With domain injection (detected: solidity):
"This is a Solidity smart contract. In addition to standard analysis, assess:
- gasComplexity: estimated gas cost pattern (low/medium/high/unbounded)
- reentrancyRisk: whether this function is vulnerable to reentrancy attacks
- storageSlotLayout: whether this modifies storage in a predictable slot pattern"
```

### Integration point (minimal wiring required)

1. Add a `detectDomain(filePaths: string[]): DetectedDomain | null` function that inspects file extensions and imports — pure static analysis, no LLM
2. Pass the detected domain to `KnowledgeGeneratorConfig`
3. In `generateForFunction()`, pass domain to `extractSemanticsWithLLM()` and add a new `extractDomainDimensionsWithLLM()` call when domain is non-null
4. Store results in `knowledge.domainDimensions`
5. Wire `projectType` (already computed in `unified_init.ts` line 433) into this pipeline

## Before/After

**Before:** Indexing a Solidity project produces `UniversalKnowledge` with empty `security.vulnerabilities[]`, generic `semantics.purpose.summary`, and zero information about gas costs, storage layout, or reentrancy — the three most critical dimensions for smart contract audit.

**After:**
```json
{
  "domainDimensions": {
    "gasComplexity": { "value": "high", "confidence": 0.82, "evidenceRefs": ["loop over unbounded array at line 45"], "domain": "solidity", "extractedBy": "llm" },
    "reentrancyRisk": { "value": true, "confidence": 0.91, "evidenceRefs": ["external call before state update at line 67"], "domain": "solidity", "extractedBy": "llm" },
    "storageSlotLayout": { "value": "predictable", "confidence": 0.75, "evidenceRefs": ["struct packing at lines 12-18"], "domain": "solidity", "extractedBy": "static" }
  }
}
```

## Acceptance Criteria

1. **Audit existing domain detection** — confirm that `COMPLETE_DOMAIN_TAXONOMY`, `domain_registry.ts` (engines), and `knowledge/registry/domain_registry.ts` are not already wired into the extraction pipeline (no assumptions allowed; evidence required)
2. **Audit `projectType` flow** — confirm that the value computed in `unified_init.ts` line 433 is not already passed to `UniversalKnowledgeGenerator` or any extractor
3. **Audit `UniversalKnowledge` for existing extension points** — confirm no `Record<string, unknown>` escape hatch exists (currently: confirmed absent)
4. **Implement `detectDomain(filePaths: string[]): DetectedDomain | null`** using static file extension + import pattern matching only (no LLM required for detection)
5. **Add `domainDimensions?: Record<string, DomainDimension>` to `UniversalKnowledge`** as an open extension point
6. **Wire domain detection into `KnowledgeGeneratorConfig`** — the generator must receive detected domain and pass it to extractors
7. **Inject domain context into `extractSemanticsWithLLM()`** prompts when domain is detected
8. **Implement domain dimension extraction for at least 2 domains** (recommended: Solidity and embedded/firmware) with evidence-backed values
9. **Wire `projectType` from `unified_init.ts`** into the knowledge generation pipeline
10. **Storage compatibility** — `domainDimensions` must round-trip through `knowledgeToRecord()` / `UniversalKnowledgeRecord` without data loss
11. **Unit tests** for domain detection signals (`.sol` → blockchain, `*.tf` → infrastructure, torch imports → ml)
12. **Integration test** against a real Solidity or embedded project fixture

## Impact

**NPS: HIGH** — Makes LiBrainian genuinely useful for non-TypeScript projects. The current domain blindness is a silent failure: engineers get results that look complete but are missing the most critical information for their domain.
**Effort: L** — Schema extension is low-risk (additive optional field). Domain detection is simple static analysis. LLM prompt injection adds one conditional branch. The main work is writing domain-specific prompt templates and evidence extraction logic.

[feature-proposal:knowledge:adaptive-dimensions]
